File: kernel/src/main/java/org/efaps/admin/ui/Menu.java
/*
 * Copyright 2003 - 2009 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Revision:        $Rev$
 * Last Changed:    $Date$
 * Last Changed By: $Author$
 */

package org.efaps.admin.ui;

import static org.efaps.admin.EFapsClassNames.MENU;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.efaps.admin.EFapsClassNames;
import org.efaps.admin.datamodel.Type;
import org.efaps.util.cache.CacheReloadException;

/**
 * @author tmo
 * @version $Id$
 * @todo description
 */
public class Menu extends AbstractMenu {

  /**
   * The static variable defines the class name in eFaps.
   */
  public final static EFapsClassNames EFAPS_CLASSNAME = MENU;

  /**
   * Logging instance used in this class.
   */
  private final static Logger LOG = LoggerFactory.getLogger(Menu.class);

  /**
   * Stores the mapping from type to tree menu.
   */
  private final static Map<Type, Menu> TYPE2MENUS = new HashMap<Type, Menu>();

  /**
   * Stores all instances of class {@link Menu}.
   *
   * @see #getCache
   */
  private static MenuCache CACHE = new MenuCache();

  /**
   * Constructor to set the id and name of the menu object.
   *
   * @param _id
   *                id of the command to set
   * @param _name
   *                name of the command to set
   */
  public Menu(final Long _id, final String _uuid, final String _name) {
    super(_id, _uuid, _name);
  }

  /**
   * An sub command or menu with the given id is added to this menu.
   *
   * @param _context
   *                eFaps context for this request
   * @param _sortId
   *                id used to sort
   * @param _id
   *                command / menu id
   * @throws CacheReloadException
   */
  @Override
  protected void add(final long _sortId, final long _id)  {
    final Command command = Command.get(_id);
    if (command == null) {
      final Menu subMenu = Menu.get(_id);
      add(_sortId, subMenu);
    } else {
      add(_sortId, command);
    }
  }

  /**
   * Sets the link properties for this object.
   *
   * @param _linkType
   *                type of the link property
   * @param _toId
   *                to id
   * @param _toType
   *                to type
   * @param _toName
   *                to name
   */
  @Override
  protected void setLinkProperty(final EFapsClassNames _linkType,
                                 final long _toId,
                                 final EFapsClassNames _toType,
                                 final String _toName)
      throws Exception
  {
    switch (_linkType) {
      case LINK_MENUISTYPETREEFOR:
        final Type type = Type.get(_toId);
        if (type == null) {
          LOG.error("Menu '"
              + getName()
              + "' could not defined as type "
              + "tree menu for type '"
              + _toName
              + "'! Type does not "
              + "exists!");
        } else {
          TYPE2MENUS.put(type, this);
        }
        break;
      default:
        super.setLinkProperty(_linkType, _toId, _toType, _toName);
    }
  }

  // ///////////////////////////////////////////////////////////////////////////

  /**
   * Returns for given parameter <i>_id</i> the instance of class {@link Menu}.
   *
   * @param _id
   *                id to search in the cache
   * @return instance of class {@link Menu}
   * @throws CacheReloadException
   * @see #getCache
   */
  static public Menu get(final long _id)  {
    return CACHE.get(_id);
  }

  /**
   * Returns for given parameter <i>_name</i> the instance of class
   * {@link Menu}.
   *
   * @param _name
   *                name to search in the cache
   * @return instance of class {@link Menu}
   * @throws CacheReloadException
   * @see #getCache
   */
  static public Menu get(final String _name)  {
    return CACHE.get(_name);
  }

  /**
   * Returns for given parameter <i>UUID</i> the instance of class {@link Menu}.
   *
   * @param _uuid
   *                UUID to search in the cache
   * @return instance of class {@link Menu}
   * @throws CacheReloadException
   * @see #getCache
   */
  static public Menu get(final UUID _uuid)  {
    return CACHE.get(_uuid);
  }

  /**
   * Returns for given type the type tree menu. If no type tree menu is defined
   * for the type, it is searched if for parent type a menu is defined.
   *
   * @param _type
   *                type for which the type tree menu is searched
   * @return type tree menu for given type if found; otherwise <code>null</code>.
   */
  public static Menu getTypeTreeMenu(final Type _type) {
    Menu ret = TYPE2MENUS.get(_type);
    if ((ret == null) && (_type.getParentType() != null)) {
      ret = getTypeTreeMenu(_type.getParentType());
    }
    return ret;
  }

  /**
   * Static getter method for the type hashtable {@link #CACHE}.
   *
   * @return value of static variable {@link #CACHE}
   */
  protected static UserInterfaceObjectCache<Menu> getCache() {
    return CACHE;
  }

  private static class MenuCache extends UserInterfaceObjectCache<Menu> {

    protected MenuCache() {
      super(Menu.class);
    }
  }
}

File: shell/src/main/resources/org/efaps/js/classes/Menu.js
/*
 * Copyright 2006 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:          tmo
 * Revision:        $Rev$
 * Last Changed:    $Date$
 * Last Changed By: $Author$
 */

importClass(Packages.org.efaps.db.Instance);
importClass(Packages.org.efaps.db.SearchQuery);

function Menu(_name)  {
  this.object = new EFapsInstance("Admin_UI_Menu", _name);
  this.instance = new Instance(Shell.getContext(), this.object.oid);

  // add helper methods for getter and setter 
  this._setAttrValue      = Abstract.prototype._setAttrValue;
  this._getAttrValue      = Abstract.prototype._getAttrValue;

  // add getter and setter methods
  this.getInstance        = Abstract.prototype.getInstance;
  this.getId              = Abstract.prototype.getId;
  this.getOid             = Abstract.prototype.getOid;
  this.getType            = Abstract.prototype.getType;
  this.getName            = Abstract.prototype.getName;
  this.getRevision        = Abstract.prototype.getRevision;
  this.setRevision        = Abstract.prototype.setRevision;

  // add property methods
  this._writeProperties   = Abstract.prototype._writeProperties;
  this.addProperty        = Abstract.prototype.addProperty;
  this.cleanupProperties  = Abstract.prototype.cleanupProperties;
  this.deleteProperty     = Abstract.prototype.addProperty;
  this.printProperties    = Abstract.prototype.printProperties;

  // add link methods
  this._createLink        = UIAbstract.prototype._createLink;
  this._writeLinks        = UIAbstract.prototype._writeLinks;
  this.cleanupLinks       = UIAbstract.prototype.cleanupLinks;
  this.addIcon            = Command.prototype.addIcon;
  this.addTargetForm      = Command.prototype.addTargetForm;
  this.addTargetMenu      = Command.prototype.addTargetMenu;
  this.addTargetSearch    = Command.prototype.addTargetSearch;
  this.addTargetTable     = Command.prototype.addTargetTable;

  // add common methods
  this._writeHeader       = Abstract.prototype._writeHeader;
}

///////////////////////////////////////////////////////////////////////////////
// instance variables

/**
 * Prefix of script file names.
 */
Menu.prototype.FILE_PREFIX = new String("MENU_");

/**
 * Name of the variable used in update scripts.
 */
Menu.prototype.VARNAME     = new String("MENU");

///////////////////////////////////////////////////////////////////////////////
// child methods

Menu.prototype.cleanupChilds = function()  {
  var query = new SearchQuery();
  query.setExpand(Shell.getContext(), this.getOid(), "Admin_UI_Menu2Command\\FromMenu");
  query.addSelect(Shell.getContext(), "OID");
  query.executeWithoutAccessCheck();
  while (query.next())  {
    var oid = query.get(Shell.getContext(), "OID");
    var del = new Delete(Shell.getContext(), oid);
    del.execute(Shell.getContext());
  }
  query.close();
}

Menu.prototype.printChilds = function() {
/*  var childs = this.object.Admin_UI_Menu2Command\FromMenu;
  if (childs!=null)  {
    var i = 0;
    while (i < childs.length) {
      print(childs[i].ToCommand.Name);
      i=i+1;
    }
  }
*/
}

/**
 * @TODO what happens if object not exists?
 */
Menu.prototype.addChild = function(_object)  {
  if (typeof _object == "object")  {
    if (_object instanceof Command || _object instanceof Menu)  {
if (_object.getId()!=null && _object.getId()!="0")  {
      var insert = new Insert(Shell.getContext(), "Admin_UI_Menu2Command");
      insert.add(Shell.getContext(), "FromMenu", this.getId());
      insert.add(Shell.getContext(), "ToCommand", _object.getId());
      insert.executeWithoutAccessCheck();
} else  {
      print("!!!!!!!!!!!!!!!!!!! object  '"+_object+"' not found!");
}
    } else  {
      print("!!!!!!!!!!!!!!!!!!! unknown object  '"+_object+"'!");
    }
  }
}

/**
 * Writes the JS update scripts for the childs.
 *
 * @param _file (Writer)  open file to write through
 */
Menu.prototype._writeChilds = function(_file)  {
  var query = new SearchQuery();
  query.setExpand(Shell.getContext(), this.getOid(), "Admin_UI_Menu2Command\\FromMenu.ToCommand");
  query.addSelect(Shell.getContext(), "Type");
  query.addSelect(Shell.getContext(), "Name");
  query.executeWithoutAccessCheck();
  while (query.next())  {
    var type = query.get(Shell.getContext(), "Type").getName();
    var name = query.get(Shell.getContext(), "Name");
    if (type=="Admin_UI_Command")  {
      _file.println("  addChild(new Command(\""+name+"\"));");
    } else  {
      _file.println("  addChild(new Menu(\""+name+"\"));");
    }
  }
  query.close();
}

///////////////////////////////////////////////////////////////////////////////
// common methods

Menu.prototype.cleanup = function()  {
  this.cleanupProperties();
  this.cleanupLinks();
  this.cleanupChilds();
}

Menu.prototype._create = function(_name)  {
  var insert = new Insert(Shell.context, "Admin_UI_Menu");
  insert.add(Shell.context, "Name", _name);
  insert.executeWithoutAccessCheck();
  this.instance = insert.getInstance();
}

Menu.prototype.update = function(_fileName, _objName)  {
  if (this.getOid()==null || this.getOid()=="" || this.getOid()=="0")  {
    print("  - create");
    this._create(_objName);
  } else  {
    print("  - cleanup");
    this.cleanup();
  }
  MENU = this;
  print("  - import");
  load(_fileName);
  MENU = null;
}

/**
 * Creates the update script for current menu.
 *
 * @param _path   (String)  file is created in this path
 * @param _author (String)  author used in the header
 */
Menu.prototype.writeUpdateScript = function(_path, _author)  {
  var file = new PrintWriter(new FileWriter(_path+"/"+this.FILE_PREFIX+this.getName()+".js"));

  this._writeHeader(file, _author);

  file.println("with ("+this.VARNAME+")  {");
  this._writeProperties(file);
  this._writeLinks(file);
  this._writeChilds(file);
  file.println("}");

  file.close();
}

///////////////////////////////////////////////////////////////////////////////
// static functions

/**
 * Extract the menu name from a file name.
 *
 * @param (String)_fileName   file name
 * @return extracted menu name
 */
function getMenuNameFromFileName(_fileName)  {
  var tmp = new String(_fileName);
  return tmp.substring(Menu.prototype.FILE_PREFIX.length, tmp.length-3);
}

/**
 * Import exact one menu.
 *
 * @param (File)_fileName   file name of the menu to import
 */
function importMenu(_fileName)  {
  var fileName = new File(_fileName);
  if (fileName.getName().startsWith(Menu.prototype.FILE_PREFIX) && fileName.getName().endsWith(".js"))  {
    var objName = getMenuNameFromFileName(fileName.getName());
    print("Import Menu '"+objName+"'");
    var impMenu = new Menu(objName);
    impMenu.update(_fileName, objName);
  }
}

/**
 * All menus in the list are created physically in the database (only the 
 * menu, no other information like properties, links, ...). This is needed, 
 * because the menus can  be referenced by other menus, commands, searches 
 * etc.
 *
 * @param _fileList array with list of files
 */
function createMenus(_fileList)  {
  print("");
  print("Create Menus");
  print("~~~~~~~~~~~~");
  for (indx in _fileList)  {
    var fileName = new File(_fileList[indx]);
    if (fileName.getName().startsWith(Menu.prototype.FILE_PREFIX) && fileName.getName().endsWith(".js"))  {
      var objName = getMenuNameFromFileName(fileName.getName());
      var menu = new Menu(objName);
      if (menu.getOid()==null || menu.getOid()=="" || menu.getOid()=="0")  {
        print("Create Menu '"+objName+"'");
        menu._create(objName);
      }
    }
  }
}

/**
 * A list of Menu Definition files is imported. The algorithmen checks first
 * for existance of a menu and creates the not existant menus.
 *
 * @param _fileList array with list of files
 */
function importMenus(_fileList)  {
  print("");
  print("Import Menus");
  print("~~~~~~~~~~~~");
  for (indx in _fileList)  {
    importMenu(_fileList[indx]);
  }
  print("");
}

/*
  var codeRev = getCodeRevision(path+"/"+fileName);
  var dbRev = MENU.object.Revision;

  if (codeRev==null || codeRev.length==0 || codeRev!=dbRev)  {
    if (codeRev!=null)  {
      MENU.object.Revision = codeRev;
      MENU.object.update();
    }
    print("  - update to revision '"+codeRev+"'");
*/

/**
 * @param _path   (String)  files are created in this path
 * @param _match  (String)  match for menu names
 * @param _author (String)  author used in the header
 */
function createScriptMenus(_path, _match, _author)  {
  var query = new Packages.org.efaps.db.SearchQuery();
  query.setQueryTypes(Shell.getContext(), "Admin_UI_Menu");
  query.addWhereExprEqValue(Shell.getContext(), "Name", _match);
  query.addSelect(Shell.getContext(), "Name");
  
  query.executeWithoutAccessCheck();
  
  print("");
  print("Create Menu Scripts:");
  print("~~~~~~~~~~~~~~~~~~~~");
  while (query.next())  {
    var name = query.get(Shell.getContext(), "Name");
    print("  - '"+  name  +"'");
    var menu = new Menu(name);
    menu.writeUpdateScript(_path, _author);
  }

  query.close();
}
